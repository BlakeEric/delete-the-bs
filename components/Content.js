const Content = (props) => (
  <div className="searchResults">
      {props.content.image && <img src={props.content.image} />}
      <h1 className="title">{props.content.title}</h1>
      <h3 className="subTitle">{props.content.description}</h3>
      <p className="pageContent" dangerouslySetInnerHTML={{ __html: props.content.text.replace(/\n/g, '<br />')}} />
      <button className="button-light endcontent-currentUrl-resetButton" onClick={props.reset}>&larr; Start Over</button>
  </div>
)

export default Content
