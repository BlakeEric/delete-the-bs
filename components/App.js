import React, { Component } from 'react';
import SearchField from './SearchField'
import Content from './Content'
import PrevSearches from './PrevSearches'
import getConfig from 'next/config'
const { publicRuntimeConfig } = getConfig()
const baseUrl = publicRuntimeConfig.baseUrl || process.env.BASE_URL;



export default class App extends Component {

 /*
  * Return an object of blank state values.
  * remountKey will be unique each time the object is regenerated
  * and will force remound of SearchField
  */
  initialState = () => {
    return {
      url: {
        value: "",
        touched: false
      },
      content: null, //object with disc, image, text, copyright
      isLoading: false,
      error: null,
      successfulSearches: []
    }
  }

  state = this.initialState();

  generateSearchList = () => {

    const url = this.state.url.value;

    const searches = [url, ...this.state.successfulSearches.filter((item, index) =>
      item !== url && index < 5
    )]

    return searches;
  }

 /*
  * Make an API request to scrape the url webpage content
  */
  submitUrl = () => {

    this.setState({
      isLoading: true,
      error: null,
      content: null,
      url: {
        ...this.state.url,
        touched: false
      }
    });

    fetch(encodeURI(`${baseUrl}/api/scrape?url=${this.state.url.value}`))
    .then(res => {
      return res.json()
    })
    .then((data) => {
      if (data.error) {
        if(data.error.statusCode === 404)
          throw "Webpage not found"

        throw "Request error"
      }

      return data
    })
    .then((data) => {
      this.setState({
        content: data,
        isLoading: false,
        successfulSearches: this.generateSearchList()
      })
    })
    .catch((err) => {
      this.setState({
        error: err,
        isLoading: false
      });
    })
  }

  /*
   * Keep track of controlled form in state
   */
   setUrl = (urlInput) => {
     this.setState({
       url: {
         value: urlInput,
         touched: true
       }
     })
   }

   /*
    * Resubmit a previous item
    */
    resubmit = (value) => {
      this.setState({
        url: {
          value: value,
        }
      },
      () => this.submitUrl()
    )}

  /*
   * Reset App to it's initial state
   */
  reset = () => {
    this.setState(this.initialState());
  }


  render() {
    return (
      <>
        <section className={`pageInner-header ${this.state.isLoading || this.state.content || this.state.error ? 'top' : ''}`}>
          <div className="pageInner-titleWrap">
            <h1 className="title">Delete the Bullshit</h1>
            <h3 className="subTitle">
              View the content of any website distraction-free
              (without ads, trackers, or other bullshit).
            </h3>
          </div>
          <SearchField key={this.state.remountKey} url={this.state.url} setUrl={this.setUrl} submitUrl={this.submitUrl} isLoading={this.state.isLoading} />
        </section>

        {this.state.content && !this.state.isLoading ?
          <Content content={this.state.content} isLoading={this.state.isLoading} url={this.state.url} reset={this.reset}/>
        : ''}

        {this.state.isLoading ?
          <div className="loadingText">
            <h2>Deleting Bullshit</h2>
          </div>
        : ''}

        {this.state.error ?
          <div className="errorMessage">
            <h2>Oops, that didn't work...</h2>
            <p>Reason: {this.state.error}</p>
            <p>Make sure you submitted a valid url or try a different one</p>
          </div>
        : ''}

        <PrevSearches handleResubmit={this.resubmit}data={this.state.successfulSearches} />
      </>
    )
  }
}
