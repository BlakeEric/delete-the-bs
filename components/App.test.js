/* eslint-env jest */
import ReactDOM from 'react-dom'
import { shallow, mount } from 'enzyme'
import toJson from 'enzyme-to-json'

import App from './App'
import Content from './Content'

let mockContent = {
  image: "https://upload.wikimedia.org/wikipedia/commons/6/6c/Campo_de_Blue_Bonnets.jpeg",
  title: "Mock Title",
  description: "Lorem ipsum dolor sit amet, consectetur adipiscing elit.",
  text: "Sed ut perspiciatis unde omnis iste natus error sit voluptatem accusantium\
  doloremque laudantium, totam rem aperiam, eaque ipsa quae ab illo inventore veritatis\
  et quasi architecto beatae vitae dicta sunt explicabo. Nemo enim ipsam voluptatem quia\
  voluptas sit aspernatur aut odit aut fugit, sed quia consequuntur magni dolores eos\
  qui ratione voluptatem sequi nesciunt.\nNeque porro quisquam est, qui dolorem ipsum quia dolor\
  sit amet, consectetur, adipisci velit, sed quia non numquam eius modi tempora incidunt\
  ut labore et dolore magnam aliquam quaerat voluptatem."
}

global.fetch = jest.fn(() => Promise.resolve({ json: () => mockContent}))

describe('App component', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<App />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('renders UI as expected', () => {
    const wrapper = mount(<App />)

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('sets state appropriately when form is submitted', async () => {

    const wrapper = mount(<App />)

    const input = wrapper.find('input');
    input.instance().value = "http://wikipedia.com";
    input.simulate('change');

    const form = wrapper.find('form').first()
    form.simulate('submit');

    expect(wrapper.state().content).toEqual(null)
    expect(wrapper.state().isLoading).toEqual(true)

    // wait until all promises are resolved
    await tick();

    expect(wrapper.state().content).toEqual(mockContent)
    expect(wrapper.state().isLoading).toEqual(false)
  });

  it('modifies ui appropriately when form is submitted', () => {

    const wrapper = mount(<App />)

    const input = wrapper.find('input');
    input.instance().value = "http://wikipedia.com";
    input.simulate('change');

    const form = wrapper.find('form').first()
    form.simulate('submit');

    expect(wrapper.find('.pageInner-header').first().hasClass('top')).toEqual(true)

  });

  it('renders <Content /> if it exists in state and resets UI when "reset" button is clicked', async() => {
    const wrapper = mount(<App />)
    const input = wrapper.find('input');
    input.instance().value = "http://wikipedia.com";
    input.simulate('change');

    wrapper.setState({content: mockContent})
    expect(wrapper.exists('Content')).toEqual(true)
    const resetButton = wrapper.find('.endcontent-currentUrl-resetButton').first()

    resetButton.simulate('click')
    wrapper.update();

    expect(wrapper.exists('Content')).toEqual(false)
    expect(wrapper.find('input').instance().value).toEqual("")
  });

  describe("prevSearches", () => {
    it('Adds previous search to state', async () => {
      const wrapper = mount(<App />)
      const input = wrapper.find('input');
      input.instance().value = "http://wikipedia.com";
      input.simulate('change');

      const form = wrapper.find('form').first()
      form.simulate('submit');

      // wait until all promises are resolved
      await tick();

      input.instance().value = "http://google.com";
      input.simulate('change');
      form.simulate('submit');

      await tick();

      expect(wrapper.state().successfulSearches).toContain('http://wikipedia.com')
      expect(wrapper.state().successfulSearches).toContain('http://google.com')
      expect(wrapper.state().successfulSearches.length).toEqual(2)

    });
  })


  it('displays an error message if form submission is NOT successful', async () => {
    global.fetch = jest.fn(() => Promise.reject( "error" ))

    const wrapper = mount(<App />)

    const input = wrapper.find('input');
    input.instance().value = "http://wikipedia.com";
    input.simulate('change');

    const form = wrapper.find('form').first()
    form.simulate('submit');

    expect(wrapper.state().error).toBeFalsy()
    expect(wrapper.exists(".errorMessage")).toBeFalsy()

    // wait until all promises are resolved
    await tick();
    wrapper.update();

    expect(wrapper.state().error).toBeTruthy()
    expect(wrapper.exists(".errorMessage")).toBeTruthy()
  });

});


// Helper function returns a promise that resolves after all other promise mocks,
// even if they are chained like Promise.resolve().then(...)
function tick() {
  return new Promise(resolve => {
    setTimeout(resolve, 0);
  })
}
