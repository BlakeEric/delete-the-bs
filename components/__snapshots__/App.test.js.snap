// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component renders UI as expected 1`] = `
<App>
  <section
    className="pageInner-header "
  >
    <div
      className="pageInner-titleWrap"
    >
      <h1
        className="title"
      >
        Delete the Bullshit
      </h1>
      <h3
        className="subTitle"
      >
        View the content of any website distraction-free (without ads, trackers, or other bullshit).
      </h3>
    </div>
    <SearchField
      isLoading={false}
      setUrl={[Function]}
      submitUrl={[Function]}
      url={
        Object {
          "touched": false,
          "value": "",
        }
      }
    >
      <div
        id="searchWrapper"
      >
        <form
          onSubmit={[Function]}
        >
          <input
            onChange={[Function]}
            placeholder="Paste a website URL to delete bullshit"
            type="text"
            value=""
          />
          <button
            className="contactForm-submitButton"
            disabled={true}
            type="submit"
          >
            Submit
          </button>
        </form>
      </div>
    </SearchField>
  </section>
  <PrevSearches
    data={Array []}
  />
</App>
`;
